source/common/grpc/google_grpc_creds_impl.cc

```diff
@@ -0,0 +1,165 @@
+#include "common/grpc/google_grpc_creds_impl.h"
+
+#include "envoy/api/v2/core/grpc_service.pb.h"
+#include "envoy/grpc/google_grpc_creds.h"
+#include "envoy/registry/registry.h"
+
+#include "common/config/datasource.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * TODO: Create GoogleGrpcCredentialsFactory for each built-in credential type defined in Google
+ * gRPC.
+ */
+
+/**
+ * Default implementation of Google Grpc Credentials Factory
+ * Uses ssl creds if available, or defaults to insecure channel.
+ */
+class DefaultGoogleGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory {
+
+public:
+  std::shared_ptr<grpc::ChannelCredentials>
+  getChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config) override {
+    const auto& google_grpc = grpc_service_config.google_grpc();
+    std::shared_ptr<grpc::ChannelCredentials> creds = grpc::InsecureChannelCredentials();
+    for (const auto& credential : google_grpc.credentials()) {
+      if (credential.has_channel_credentials() &&
+          credential.channel_credentials().has_ssl_credentials()) {
+        const grpc::SslCredentialsOptions ssl_creds = {
+            .pem_root_certs = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().root_certs(), true),
+            .pem_private_key = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().private_key(), true),
+            .pem_cert_chain = Config::DataSource::read(
+                credential.channel_credentials().ssl_credentials().cert_chain(), true),
+        };
+        return grpc::SslCredentials(ssl_creds);
+      }
+    }
+    return creds;
+  }
+
+  std::string name() const override { return "envoy.grpc_credentials.default"; }
+};
+
+/**
+ * Access token implementation of Google Grpc Credentials Factory
+ * This implementation uses ssl creds for the grpc channel if available, similar to the default
+ * implementation. Additionally, it uses MetadataCredentialsFromPlugin to add a static secret to a
+ * header for call credentials. This implementation does the same thing as AccessTokenCredentials,
+ * but it's implemented as a plugin to show how a custom implementation would be created.
+ *
+ * This implementation uses the access_token field in the config to get the secret to add to the
+ * header.
+ *
+ * This can be used as an example for how to implement a more complicated custom call credentials
+ * implementation. Any blocking calls should be performed in the
+ * MetadataCredentialsFromPlugin::GetMetadata to ensure that the main thread is not blocked while
+ * initializing the channel.
+ */
+class AccessTokenGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory {
```

Comment:
We should move this under `source/extensions/grpc_credentials` and rename to reflect this is an example only (since the Google gRPC client can do this already). @mattklein123 do we have a policy on how example extensions are to be treated in the repo?